### ISSUE TITLE
[FEATURE] Run test in random order

### ISSUE THREAD

**Issue by fasenderos** [tapjs/tapjs#972]
In order to improve the quality of user tests, it would be great to run tests in a random order based on a random seed to ensure that the tests are properly isolated from each other.

The use of a seed will allow to re-run tests in the same order to reproduce a failed case for a specific order.

Other test suites have this feature; for example, you can refer to jest [randomize](https://jestjs.io/docs/cli#--randomize) or vitest [shuffle](https://vitest.dev/config/#sequence-shuffle)

**Comment by isaacs** [comment]
Do you mean the suite files, or the tests within a given file?

Tap doesn't defer your tests or put them in any sort of queue to be run at some later time, so "shuffling" is sort of a weird thing to consider. Tap just runs your test functions right away, unless it's already in the middle of something else.

For example:

```js
import t from 'tap'
console.error('a')
t.test('b', t => {
  console.error('in b')
  t.end()
})
console.error('c')
t.test('d', t => {
  console.error('in d')
  t.end()
})
console.error('e')
```

Output:

```tap
a
TAP version 14
in b
# Subtest: b
    1..0
ok 1 - b # time=0.781ms

c
in d
# Subtest: d
    1..0
ok 2 - d # time=0.336ms

e
1..2
# { total: 2, pass: 2 }
# time=9.289ms
```

As you can see, it doesn't even know about the `d` subtest until the `b` subtest is completely finished and resolved. This determinism is baked into tap, so that definitely could not be changed by default, maybe as opt-in, and only if the tests are not sync.

If they're async (and maybe whatever shuffle opt-in _forces_ them to be deferred and async), they could theoretically be shuffled. Today, if you set `t.jobs` to a number > 1, then it will run them in parallel, which shuffles them up a bit, but not quite what you're looking for, I think.

It's really not required for the tests within a given file to be isolated from one another, and it's pretty common for them _not_ to be. For example, you might be testing specifically that a given _series_ of actions has some effect, and using subtests as a sort of flow-control-plus-assertions is a handy way to do that.

**Comment by isaacs** [comment]
Approved feature: `--shuffle` flag will run test _files_ in a random order, and print the seed that was used, which can also be set via `--shuffle-seed=<number>`. It won't have any effect on the contents of each individual test file, though.

**Comment by isaacs** [comment]
Hm, actually, a `--shuffle-seed` doesn't really make tests reproducible in that way, because of the overlap resulting from parallelism. So, unless you _also_ do `--jobs=1`, it's going to be somewhat nondeterministic, since the precise ordering will be resulting from the specific timing, CPU load, etc.

Therefore, I think it's best to just add `--shuffle` to make it randomly arranged, and be done with it.
